### ISSUE TITLE
`fetchRecordSets` does not return all records in case of pagination (`nextLink`)

### ISSUE THREAD

**Issue by run1et** [StackExchange/dnscontrol#770]
I created a zone file with ~1000 records (need to migrate a pretty large zone from BIND to Azure DNS).
All these records were created on Azure DNS, but when I tried to repeat the same command, I received a set from 30 records that should be created one more time. Expected result — 0 changes. At the same time, all these records are present on Azure DNS. 

```bash
$ dnscontrol push
#266: CREATE CNAME record-cname-499.companydomain123456.com apple.com. ttl=300
SUCCESS!
#358: CREATE CNAME record-cname-500.companydomain123456.com apple.com. ttl=300
SUCCESS!
...
Done. 500 corrections.

$ dnscontrol push
----- DNS Provider: azuredns...2 corrections
#1: CREATE CNAME record-cname-499.companydomain123456.com apple.com. ttl=300
SUCCESS!
#2: CREATE CNAME record-cname-500.companydomain123456.com apple.com. ttl=300
SUCCESS!
----- Registrar: none...0 corrections
Done. 2 corrections.
```

I realized that after some amount of records (generally speaking after some amount of text in API response from Azure DNS) API paginates a result with `nextLink`:

```json
{"nextLink":"https:\/\/management.azure.com:443\/subscriptions..","value":[{"id":"\/subscriptions\/..."...}]}
```

And `fetchRecordSets` (SDK call `a.recordsClient.ListAllByDNSZoneComplete`) doesn't process this `nextLink`.

Tried to use the latest `azure-sdk-for-go` (v43.3.0 instead of v42.2.0) - the same. Branch: `master`, commit hash `ff8ce26cee5688f745f2f53c8c7654655a0e015e` 

Tried to fix that with some code like this — doesn't work, got an endless loop (6 seconds and a deferred cancel).

```golang
for recordsIterator, err := a.recordsClient.ListAllByDNSZone(ctx, *a.resourceGroup, zoneName, to.Int32Ptr(1000), ""); recordsIterator.NotDone(); err = recordsIterator.NextWithContext(ctx) {
        recordsResult := recordsIterator.Response()
        for _, r := range *recordsResult.Value {
            record := r
            records = append(records, &record)
        }
}
```

Checked some similar issue https://github.com/terraform-providers/terraform-provider-azurerm/pull/7036/files, doesn't work for me.

For this reason I changed a visibility of `a.recordsClient.ListByDNSZoneNextResults` from azure-sdk-for-go (it's private) and implemeted this function:

```golang
func (a *azureDnsProvider) fetchRecordSets(zoneName string) ([]*adns.RecordSet, error) {
	if zoneName == "" {
		return nil, nil
	}
	var records []*adns.RecordSet
	ctx, cancel := context.WithTimeout(context.Background(), 6000*time.Second)
	defer cancel()
	recordsIterator, recordsErr := a.recordsClient.ListAllByDNSZoneComplete(ctx, *a.resourceGroup, zoneName, to.Int32Ptr(1000), "")
	if recordsErr != nil {
		return nil, recordsErr
	}
	recordsResult := recordsIterator.Response()

	for _, r := range *recordsResult.Value {
		record := r
		records = append(records, &record)
	}

    // Process nextLink data
	for recordsResult.NextLink != nil {
		var err error
		recordsResult, err = a.recordsClient.ListByDNSZoneNextResults(ctx, recordsResult)
		if err != nil {
			fmt.Errorf("fetchRecordSets NextLink. loading Resource List: %+v", err)
			continue
		}

		for _, r := range *recordsResult.Value {
			fmt.Printf("fetchRecordSets NextLink FQDN: [%v]\n", *r.Fqdn)
			record := r
			records = append(records, &record)
		}
	}

	return records, nil
}
```

And this variant is working perfectly. But I can't use it because of needed changes in azure-sdk-for-go.

Probably you can recommend how to use azure-sdk-for-go and activate `result.fn` for fetching all records.

```golang
func (client RecordSetsClient) ListAllByDNSZone
...
result.fn = client.listAllByDNSZoneNextResults
...
```

**Comment by run1et** [comment]
Forgot to add a test zone file:
[dnsconfig.txt](https://github.com/StackExchange/dnscontrol/files/4804447/dnsconfig.txt)

**Comment by tlimoncelli** [comment]
Hi there!

You've done an excellent analysis so far! 

I've been able to reproduce the issue.  I've updated integration tests to make it easier to test this situation:

```
$ export AZURE_SUBSCRIPTION_ID='redacted'
$ export AZURE_RESOURCE_GROUP='redacted'
$ export AZURE_TENANT_ID='redacted'
$ export AZURE_CLIENT_ID='redacted'
$ export AZURE_CLIENT_SECRET='redacted'
$ export AZURE_DOMAIN='your-domain-here.com'
$ go test -v -verbose -provider AZURE_DNS -run TestDNSProviders -start 15 -end 15
```

(note: this is on the master branch)

I agree that the go module should either return all records (hide that there is any paging at all) or expose the paging and provide a way to get specific pages.  I'd prefer it would hide the paging from the caller.

Could you file a report on https://github.com/Azure/azure-sdk-for-go/issues ?

Tom

**Comment by run1et** [comment]
Thanks a lot, Tom! I will create an issue in Azure/azure-sdk-for-go and add a link here.

Oleg.

**Comment by meowingtons** [comment]
Oleg,

Did you ever create an issue for this in the Azure/azure-sdk-for-go project? We're running into this issue as well.

Alex

**Comment by run1et** [comment]
Hi guys. Thank you for reminding, I've created a corresponding issue in azure-sdk-for-go:
https://github.com/Azure/azure-sdk-for-go/issues/11763

**Comment by run1et** [comment]
"It looks like the fix for #9510 was merged":
- https://github.com/Azure/azure-sdk-for-go/issues/11763
- https://github.com/Azure/azure-sdk-for-go/issues/9510

(I am "a seasonal" Golang "user") Tried to use `azure-sdk-for-go`
- `master`: v44.2.0+incompatible
- `track2`: v40.6.1-0.20200724172808-9ff7e6f362f0+incompatible

Both variants didn't fix this problem. Need to dive into the code.

**Comment by run1et** [comment]
I've fixed this issue in #791
